substrate-node/pallets/pallet-tfgrid/src/migrations/v14.rs

```diff
@@ -0,0 +1,313 @@
+use crate::{
+    ip::{GW4, GW6, IP4, IP6},
+    types,
+    types::StorageVersion,
+    Config, Error, FarmInfoOf, Farms, InterfaceOf, LocationOf, Nodes, PalletVersion, PubConfigOf,
+    PublicIpOf, SerialNumberOf, TFGRID_NODE_VERSION,
+};
+use frame_support::{
+    pallet_prelude::Weight, traits::ConstU32, traits::Get, traits::OnRuntimeUpgrade, BoundedVec,
+};
+use log::{debug, info};
+use sp_std::marker::PhantomData;
+use tfchain_support::{
+    traits::PublicIpModifier,
+    types::{Farm, Node},
+};
+
+use valip::ip4::{Ip as IPv4, CIDR as IPv4Cidr};
+use valip::ip6::{Ip as IPv6, CIDR as IPv6Cidr};
+
+#[cfg(feature = "try-runtime")]
+use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+
+pub struct FixPublicIP<T: Config>(PhantomData<T>);
+
+impl<T: Config> OnRuntimeUpgrade for FixPublicIP<T>
+where
+    IP4<T>: From<<T as Config>::IP4>,
+    GW4<T>: From<<T as Config>::GW4>,
+    IP6<T>: From<<T as Config>::IP6>,
+    GW6<T>: From<<T as Config>::GW6>,
+{
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<(), &'static str> {
+        assert!(PalletVersion::<T>::get() <= types::StorageVersion::V13Struct);
```

Comment:
Here is the thing I noticed: all the pre_upgrade checks are performed before the migrations happen. 

So for example; if you go from v1 to v3 in 2 migrations, the pre_upgrade check for v3 would fail since the v2 migration didnt run yet.

Reply:
I get the logic
So it means the `pre_upgrade()` version check for v10 v11 v12 are not good since we have:
- V10:
```rust
    fn pre_upgrade() -> Result<(), &'static str> {
        assert!(PalletVersion::<T>::get() >= types::StorageVersion::V9Struct);
```
- V11:
```rust
    fn pre_upgrade() -> Result<(), &'static str> {
        assert!(PalletVersion::<T>::get() >= types::StorageVersion::V10Struct);
```
- V12:
```rust
    fn pre_upgrade() -> Result<(), &'static str> {
        assert!(PalletVersion::<T>::get() >= types::StorageVersion::V11Struct);
```

Reply:
Well, it is good but it wont work for networks (devnet) that are on V12. Since these assertion will fail.